### PR TITLE
[#74] Implement 'main' function

### PR DESCRIPTION
**Pull request by krzema12**
Both argumentless and "argumentful" function is supported.

# Testing

Tried to run the generated output, and t

he control flow correctly goes to `main`.

Here's how it works for argumentless function:
Screenshot from 2022-02-18 10-16-25

### CODE REVIEW

File: compiler/ir/backend.py/src/org/jetbrains/kotlin/ir/backend/py/transformers/irToPy/IrModuleToPyTransformer.kt [hunk 1]
```diff
@@ -84,20 +101,75 @@ class IrModuleToPyTransformer(
...
+        val callToMain = generateCallToMain(modules)
+        val imports = generateImports()
+
         val moduleBody = generateModuleBody(modules, rootContext)
         val program = Module(
             body = moduleBody,
             type_ignores = emptyList(),
-        )
+        ).let {
+            it.copy(body = imports + it.body + callToMain)
```
Comment (SerVB) [inline compiler/ir/backend.py/src/org/jetbrains/kotlin/ir/backend/py/transformers/irToPy/IrModuleToPyTransformer.kt:112]: We can actually import inside `callToMain`. Would be a bit more compact both in our code and in generated code. Of course it won't be clean code, but I guess we aren't trying to generate clean code for now.

But I'm not insisting.

File: python/box.tests/test/org/jetbrains/kotlin/python/test/BasicIrBoxTest.kt [hunk 2]
```diff
@@ -84,6 +85,7 @@ abstract class BasicIrBoxTest(
                     testPackage = testFactory.testPackage,
                     needsFullIrRuntime = needsFullIrRuntime,
                     isMainModule = isMainModule,
+                    mainCallParameters = MainCallParameters.noCall(),
```
Comment (SerVB) [inline python/box.tests/test/org/jetbrains/kotlin/python/test/BasicIrBoxTest.kt]: If this is a constant, probably we can inline it where it's finally used (for example, line 287) to simplify the file? At least that is what I try to do with this file :)

**Review by SerVB — APPROVED** [review]
Cool! A couple of minor comments

File: compiler/ir/backend.py/src/org/jetbrains/kotlin/ir/backend/py/transformers/irToPy/IrModuleToPyTransformer.kt (hunk 1 above)
Comment (krzema12) [reply compiler/ir/backend.py/src/org/jetbrains/kotlin/ir/backend/py/transformers/irToPy/IrModuleToPyTransformer.kt:112]: I'm rather resistant to importing stuff in other place than the beginning of code. It's some unwritten convention that I sew being broken only in very peculiar cases (e.g. in Python, conditional importing guarded by `if`, depending on Python version). Let me pass on that :)

File: python/box.tests/test/org/jetbrains/kotlin/python/test/BasicIrBoxTest.kt (hunk 2 above)
Comment (krzema12) [reply python/box.tests/test/org/jetbrains/kotlin/python/test/BasicIrBoxTest.kt]: Yeah, I was struggling to recreate the whole structure for `mainCallParameters` and ended up with something unclean. Let me simplify it like you propose, and we can always parameterize if needed.

**Comment by SerVB** [comment]
Should we also now close #74? Or maybe add a test for it before closing?

**Comment by krzema12** [comment]
@SerVB yeah, I'm thinking of an e2e test, then I'll resolve it.

**Pull request merged**